---
title: "Chromosome-specific methylation analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE, warning=FALSE, results=FALSE}
#load packages, source functions
library(Rsamtools)
library(tidyverse)
library(slider)
library(ggplot2)
library(grid)
library(gridExtra)
source("GC_p_meth_45s.R")
```

```{r}
#read in data generated by 45_s_parse.R
read_sums <- read_tsv("/path/to/read_summaries.txt") %>% mutate(chr = as.factor(chr), reg = as.factor(reg))
head(read_sums)
```
Plots

```{r}
read_sums %>% ggplot(aes(x = chr, y = mean, fill = chr)) + geom_violin() + geom_boxplot(width=0.1)
```

```{r}
read_sums %>% filter(length > 200) %>% ggplot(aes(x = chr, y = mean, fill = reg)) + geom_violin() 
```

```{r}
read_sums %>% filter(length > 200) %>% ggplot() + geom_histogram(aes(x = mean), fill = "lightblue") + facet_wrap(~chr)
```

```{r}
read_sums %>% group_by(chr, reg) %>% summarize(n = n()) %>% 
  ggplot() + geom_bar(mapping = aes(x = chr, y = n, fill = reg), 
                      stat = "identity", position=position_dodge()) + 
  xlab("Number of reads mapping to region") + ylab("chromosome") + 
  ggtitle("Number of reads mapping to each acro p-arm region by chromosome") + 
  geom_text(aes(x = chr, y = n, fill = reg, label=n), position=position_dodge(.9), vjust = 0, size=3.5)

```
